SessionPlayer/SessionPlayer.js

```diff
@@ -121,76 +174,118 @@ export default class SessionPlayer extends Component {
-            currentInstruction,
-            currentInstructionIndex,
             guideState,
+            currentGuide
         } = this.state;
 
-        let DisplayedInstruction;
-
         switch (guideState) {
             case 'playing':
-                DisplayedInstruction = Guides[currentGuide.guideName].Playing;
-                break;
+                return DisplayedInstruction =
+                    Guides[currentGuide.guideName].Playing;
             case 'paused':
-                DisplayedInstruction = Guides.Global.Paused;
+                return DisplayedInstruction =
+                    Guides.Global.Paused;
             case 'finished':
-                DisplayedInstruction = Guides[currentGuide.guideName].Finished;
+                return DisplayedInstruction =
+                    Guides[currentGuide.guideName].Finished;
+            case 'all_finished':
+                return DisplayedInstruction =
+                    Guides.Global.AllFinished;
         }
+    }
+
+    render() {
+        const {
+            currentGuide,
+            currentGuideIndex,
+            currentInstruction,
+            currentInstructionIndex,
+            guideState,
+        } = this.state;
+
+        const DisplayedInstruction = this.determineDisplayedInstruction();
+        const totalGuides = Object.keys(Guides).length - 1; // -1 to remove "Global" key
 
         return (
-            <View style={styles.container}>
-                <View style={styles.sceneProcessor}>
-                    <View style={styles.scenePlayground}>
+            <SafeAreaView style={styles.container}>
+                <View style={styles.sessionPlayer}>
+                    <View style={styles.HUDContainer}>
+                        <HUD
+                            key={currentGuideIndex.toString()}
+                            guideName={currentGuide.guideName}
+                            guideState={guideState}
+                            totalGuides={totalGuides}
+                            currentGuideNumber={currentGuideIndex + 1}
+                        />
+                    </View>
+                    <View style={styles.displayedInstruction}>
                         <DisplayedInstruction
-                            key={currentGuideIndex.toString() + currentInstructionIndex.toString()}
+                            key={
+                                currentGuideIndex.toString()
+                                + currentInstructionIndex.toString()
+                                + this.state.guideState
+                            }
                             onInstructionEnd={this.onInstructionEndHandler}
                             duration={currentInstruction.duration}
                             instructionsText={currentInstruction.instructionsText}
                         />
                     </View>
                     <View style={styles.actionButtonContainer}>
+                        <Icon
```

Comment:
This long `render` makes a lot of sense. It is putting right components and correct props together, however the `<Icon>` construction involves sizing, icon name and etc. I don't think render needs to know about all of it. Something like `renderDecrementGuideButton` should be more appropriate IMO.